### PR TITLE
TICK-360 Add ModelSCCS

### PR DESCRIPTION
**Pull request by MaryanMorel**

### CODE REVIEW

File: tick/optim/model/src/sccs.cpp [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    max_sq_norm = 0;
+    ulong max_interval = get_max_interval(sample);
+    for (ulong t=0; t < max_interval; t++) {
+      row = get_longitudinal_features(sample, t);
+      // Lipschitz constant = 0 if Y_{sample, t} = 0
+      if (get_longitudinal_label(sample, t) > 0) {
+        for (ulong k = 0; k < max_interval; k++) {
+          other_row = get_longitudinal_features(sample, k);
+          sq_norm = 0;
+          for (ulong feature = 0; feature < n_lagged_features; feature++) {
```
Comment (stephanegaiffas) [inline tick/optim/model/src/sccs.cpp]: array have a norm_sq method or something like that. If you want it on a part of the array, you can use a view (makes code simpler)

File: tick/optim/model/tests/sccs_test.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+                  np.array([[1, 1],
+                            [1, 0]], dtype="float64")
+                 ]
+        self.y = [np.array([1, 0], dtype="int32"),
+                  np.array([0, 1], dtype="int32")
+                 ]
+
+    def test_loss(self):
+        """Test longitudinal multinomial model loss."""
+        # TODO: make this more complex
```
Comment (stephanegaiffas) [inline tick/optim/model/tests/sccs_test.py]: Check what we do in other Models unitests if necessary. 
But it seems that's enough tests for now right ?

File: tick/preprocessing/longitudinal_features_lagger.py [hunk 3]
```diff
@@ -144,21 +144,21 @@ def fit(self, X, censoring):
...
 
-    def transform(self, X, censoring):
+    def transform(self, X, censoring=None):
         """Add lagged features to the given features matrices list.
 
         Parameters
         ----------
         X : list of numpy.ndarray or list of scipy.sparse.csr_matrix,
-            list of length n_samples, each element of the list of 
+            list of length n_samples, each element of the list of
```
Comment (stephanegaiffas) [inline tick/preprocessing/longitudinal_features_lagger.py]: put types between `` `` `` `` :) :)

File: tick/optim/model/sccs.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+            censoring = np.full(self.n_samples, self.n_intervals,
+                                dtype="uint64")
+        censoring = check_censoring_consistency(censoring, self.n_samples)
+        features = check_longitudinal_features_consistency(features,
+                                                           base_shape,
+                                                           "float64")
+        labels = check_longitudinal_features_consistency(labels,
+                                                         (self.n_intervals,),
+                                                         "int32")
+
```
Comment (MaryanMorel) [inline tick/optim/model/sccs.py]: What do you mean ?

File: tick/optim/model/sccs.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    def _grad(self, coeffs: np.ndarray, out: np.ndarray) -> None:
+        self._model.grad(coeffs, out)
+
+    def _loss(self, coeffs: np.ndarray) -> float:
+        return self._model.loss(coeffs)
+
+    def _get_n_coeffs(self):
+        return self._model.get_n_coeffs()
+
+    def _get_lip_best(self):
```
Comment (MaryanMorel) [inline tick/optim/model/sccs.py]: Yes that's right !

File: tick/optim/model/src/sccs.cpp (hunk 1 above)
Comment (MaryanMorel) [inline tick/optim/model/src/sccs.cpp]: I don't see how using this method in this case, as I'm looking for the (row - other_row).norm_sq(). 
I could do something like `sq_norm = other_row.mult_incr(row, -1).norm_sq();`, but I'm working with BaseArray and not Array. so I don't have access to mult_incr or similar methods.

**Comment by stephanegaiffas** [comment]
Do it by hand in the model, we'll add the missing methods in arrays later :)

Le 20 juin 2017 à 17:46, Maryan <notifications@github.com> a écrit :

@MaryanMorel commented on this pull request.

In tick/optim/model/src/sccs.cpp:

> +  lip_consts.init_to_zero();
+  BaseArrayDouble row, other_row;
+
+  double max_sq_norm, sq_norm;
+
+  for (ulong sample = 0; sample < n_samples; sample++) {
+    max_sq_norm = 0;
+    ulong max_interval = get_max_interval(sample);
+    for (ulong t=0; t < max_interval; t++) {
+      row = get_longitudinal_features(sample, t);
+      // Lipschitz constant = 0 if Y_{sample, t} = 0
+      if (get_longitudinal_label(sample, t) > 0) {
+        for (ulong k = 0; k < max_interval; k++) {
+          other_row = get_longitudinal_features(sample, k);
+          sq_norm = 0;
+          for (ulong feature = 0; feature < n_lagged_features; feature++) {
I don't see how using this method in this case, as I'm looking for the (row - other_row).norm_sq().
I could do something like sq_norm = other_row.mult_incr(row, -1).norm_sq();, but I'm working with BaseArray and not Array. so I don't have access to mult_incr or similar methods.

**Comment by MaryanMorel** [comment]
@stephanegaiffas Could you get a look at the changes ?

File: tick/optim/model/src/sccs.cpp [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+                                     ArrayDouble &out) {
+  out.init_to_zero();
+  ArrayDouble inner_prod(n_intervals);
+  ArrayDouble buffer(n_lagged_features);
+  buffer.init_to_zero();
+  ulong max_interval = get_max_interval(i);
+
+  for (ulong t = 0; t < max_interval; t++)
+    inner_prod[t] = get_inner_prod(i, t, coeffs);
+  for (ulong t = max_interval; t < n_intervals; t++)
```
Comment (MaryanMorel) [inline tick/optim/model/src/sccs.cpp]: done

File: tick/optim/model/sccs.py (hunk 5 above)
Comment (MaryanMorel) [inline tick/optim/model/sccs.py]: done

File: tick/optim/model/sccs.py [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+    }
+
+    def __init__(self, n_intervals: int, n_lags: int):
+        if n_lags >= n_intervals:
+            raise ValueError("n_lags should be < n_intervals")
+
+        ModelFirstOrder.__init__(self)
+        ModelLipschitz.__init__(self)
+        self.n_lags = n_lags
+        self.n_intervals = n_intervals
```
Comment (MaryanMorel) [inline tick/optim/model/sccs.py]: done

File: tick/optim/model/src/sccs.cpp (hunk 1 above)
Comment (stephanegaiffas) [inline tick/optim/model/src/sccs.cpp]: I don't understand why you have a problem, why we don't for other models... let's talk about that ASAP

File: tick/preprocessing/longitudinal_features_lagger.py [hunk 8]
```diff
@@ -144,21 +144,21 @@ def fit(self, X, censoring):
...
-        X : list of numpy.ndarray or list of scipy.sparse.csr_matrix,
-            list of length n_samples, each element of the list of 
+        X : `list` of `numpy.ndarray` or `list` of `scipy.sparse.csr_matrix`,
+            list of length n_samples, each element of the list of
             shape=(n_intervals, n_features)
             The list of features matrices.
 
-        censoring : `numpy.ndarray`, shape=(n_samples, 1), dtype="uint64"
-            The censoring data. This array should contain integers in 
+        censoring : (optional) `numpy.ndarray`, shape=(n_samples, 1), dtype='uint64'
```
Comment (stephanegaiffas) [inline tick/preprocessing/longitudinal_features_lagger.py]: `numpy.ndarray`, shape=(n_samples, 1), dtype='uint64', default='None'
(vire le (optional))

File: tick/preprocessing/longitudinal_features_lagger.py [hunk 9]
```diff
@@ -182,22 +185,22 @@ def transform(self, X, censoring):
...
-            list of length n_samples, each element of the list of 
+        X : `list` of `numpy.ndarray` or `list` of `scipy.sparse.csr_matrix`,
+            list of length n_samples, each element of the list of
             shape=(n_intervals, n_features)
             The list of features matrices.
-            
-        censoring : `numpy.ndarray`, shape=(n_samples, 1), dtype="uint64"
-            The censoring data. This array should contain integers in 
+
+        censoring : (optional) `numpy.ndarray`, shape=(n_samples, 1), dtype="uint64"
```
Comment (stephanegaiffas) [inline tick/preprocessing/longitudinal_features_lagger.py]: cf au dessus

**Pull request merged**